File: archivos-otros-escenarios/escenario-3/vim2/py/producer/billing.service.py
import pika, json, time, random
import sys

# Wait for RabbitMQ to be ready with retry logic
max_retries = 30
retry_count = 0

while retry_count < max_retries:
    try:
        print(f"Attempting to connect to RabbitMQ (attempt {retry_count + 1}/{max_retries})")
        connection = pika.BlockingConnection(pika.ConnectionParameters(
            host='10.0.0.9',
            port=5672,
            credentials=pika.PlainCredentials('admin', 'admin123')
        ))
        print("Successfully connected to RabbitMQ!")
        break
    except pika.exceptions.AMQPConnectionError:
        retry_count += 1
        if retry_count >= max_retries:
            print("Failed to connect to RabbitMQ after maximum retries")
            sys.exit(1)
        print(f"Connection failed, retrying in 5 seconds...")
        time.sleep(5)

channel = connection.channel()

# Declare queue directly (simple approach, no exchange)
queue = 'billing'
channel.queue_declare(queue=queue, durable=True)

# Sample billing data
customers = ["Laura", "Yeison", "Santiago", "María", "Carlos"]
items = ["Capuchino", "Latte", "Espresso doble", "Mocha", "Americano"]

print("[Billing Service] Generating invoices...")
for i in range(4):
    invoice_id = f"INV-{1001 + i}"
    customer = random.choice(customers)
    item = random.choice(items)
    quantity = random.randint(1, 3)
    unit_price = random.uniform(2.5, 5.5)
    total = round(quantity * unit_price, 2)
    
    billing = {
        "invoice_id": invoice_id,
        "customer": customer,
        "item": item,
        "quantity": quantity,
        "unit_price": round(unit_price, 2),
        "total": total,
        "status": "pending",
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ")
    }
    msg = json.dumps(billing)
    
    # Send to main queue for consumers
    channel.basic_publish(
        exchange='',
        routing_key=queue,
        body=msg,
        properties=pika.BasicProperties(
            delivery_mode=2,  # make message persistent
        )
    )
    
    # Also send to monitor queue for API
    monitor_queue = f"{queue}_monitor"
    channel.queue_declare(queue=monitor_queue, durable=True)
    channel.basic_publish(
        exchange='',
        routing_key=monitor_queue,
        body=msg,
        properties=pika.BasicProperties(
            delivery_mode=2,
        )
    )
    
    print(f"[Billing Service] Factura generada: {invoice_id} - {customer} - ${total}")
    time.sleep(1.5)  # Small delay between messages

print("[Billing Service] All invoices sent!")
connection.close()

File: archivos-otros-escenarios/escenario-2/vim2/js/consumer/billing.service.js
import amqp from "amqplib";

const RABBIT_URL = "amqp://admin:admin123@10.0.0.9:5672";

const connectWithRetry = async (maxRetries = 30) => {
    for (let i = 0; i < maxRetries; i++) {
        try {
            console.log(`Attempting to connect to RabbitMQ (attempt ${i + 1}/${maxRetries})`);
            const conn = await amqp.connect(RABBIT_URL);
            
            conn.on('error', (err) => {
                console.error('Connection error:', err);
            });
            
            conn.on('close', () => {
                console.log('Connection closed');
            });
            
            console.log("Successfully connected to RabbitMQ!");
            return conn;
        } catch (error) {
            console.log(`Connection failed, retrying in 5 seconds...`);
            if (i === maxRetries - 1) {
                throw new Error("Failed to connect to RabbitMQ after maximum retries");
            }
            await new Promise(resolve => setTimeout(resolve, 5000));
        }
    }
};

const processMessage = (msg, ch) => {
    try {
        if (msg) {
            const billing = JSON.parse(msg.content.toString());
            console.log(`[Billing Service] Procesando factura ${billing.invoice_id} - Cliente: ${billing.customer} - Total: $${billing.total}`);
            ch.ack(msg);
        }
    } catch (error) {
        console.error('Error processing message:', error);
        ch.nack(msg, false, false);
    }
};

const startService = async () => {
    let conn = null;
    let ch = null;

    const setupConsumer = async () => {
        try {
            conn = await connectWithRetry();
            ch = await conn.createChannel();
            
            const queue = "billing";
            
            await ch.assertQueue(queue, { durable: true });
            ch.prefetch(1);
            
            console.log("[Billing Service] Esperando facturas...");
            
            ch.consume(queue, (msg) => processMessage(msg, ch));
            
            ch.on('error', async (err) => {
                console.error('Channel error:', err);
                await reconnect();
            });
            
        } catch (error) {
            console.error('Setup error:', error);
            await reconnect();
        }
    };

    const reconnect = async () => {
        console.log('Attempting to reconnect...');
        
        try {
            if (ch) await ch.close();
            if (conn) await conn.close();
        } catch (closeError) {
            console.error('Error closing connections:', closeError);
        }
        
        await new Promise(resolve => setTimeout(resolve, 5000));
        await setupConsumer();
    };

    await setupConsumer();
};

// Start the service
startService().catch(console.error);

// Keep process alive
process.on('SIGINT', async () => {
    console.log('\nShutting down gracefully...');
    process.exit(0);
});
